packages/core/src/locales/fi.ts

```diff
@@ -0,0 +1,197 @@
+  return Sizable[origin] ?? null;
+}
+
+const TypeNames: { [k in $ZodTypeDef["type"] | (string & {})]?: string } = {
+  string: "merkkijono",
+  number: "luku",
+  boolean: "totuusarvo",
+  bigint: "suuri kokonaisluku",
+  symbol: "symboli",
+  null: "tyhjä",
+  undefined: "määrittelemätön",
+  date: "päivämäärä",
+  object: "objekti",
+  file: "tiedosto",
+  array: "lista",
+  map: "hakemisto",
+  set: "joukko",
+  nan: "epäluku",
+  promise: "lupaus",
+};
+
+function getTypeName(type: string): string {
+  return TypeNames[type] ?? type;
+}
+
+export const parsedType = (data: any): string => {
+  const t = typeof data;
+
+  switch (t) {
+    case "number": {
+      return Number.isNaN(data) ? "epäluku" : "luku";
+    }
+    case "bigint": {
+      return Number.isNaN(data) ? "epäluku" : "suuri kokonaisluku";
+    }
+    case "boolean": {
+      return "totuusarvo";
+    }
+    case "symbol": {
+      return "symboli";
+    }
+    case "function": {
+      return "funktio";
+    }
+    case "string": {
+      return "merkkijono";
+    }
+    case "undefined": {
+      return "määrittelemätön";
+    }
+    case "object": {
+      if (Array.isArray(data)) {
+        return "lista";
+      }
+      if (data === null) {
+        return "tyhjä";
+      }
+      if (data instanceof Date) {
+        return "päivämäärä";
+      }
+      if (data instanceof Map) {
+        return "hakemisto";
+      }
+      if (data instanceof Set) {
+        return "joukko";
+      }
+      if (data instanceof File) {
+        return "tiedosto";
+      }
+      if (data instanceof Promise) {
+        return "lupaus";
+      }
+
+      if (Object.getPrototypeOf(data) !== Object.prototype && data.constructor) {
+        return data.constructor.name;
+      }
+
+      return "objekti";
+    }
+    default: {
+      return t;
+    }
+  }
+};
+
+const Nouns: {
+  [k in $ZodStringFormats | (string & {})]?: string;
+} = {
+  regex: "regex-lauseke",
+  email: "sähköpostiosoite",
+  url: "URL-osoite",
+  emoji: "emoji",
+  uuid: "UUID",
+  uuidv4: "UUIDv4",
+  uuidv6: "UUIDv6",
+  nanoid: "nanoid",
+  guid: "GUID",
+  cuid: "cuid",
+  cuid2: "cuid2",
+  ulid: "ULID",
+  xid: "XID",
+  ksuid: "KSUID",
+  datetime: "ISO-aikaleima",
+  date: "ISO-päivämäärä",
+  time: "ISO-aika",
+  duration: "ISO-kesto",
+  ipv4: "IPv4-osoite",
+  ipv6: "IPv6-osoite",
+  cidrv4: "IPv4-alue",
+  cidrv6: "IPv6-alue",
+  base64: "base64-koodattu merkkijono",
+  base64url: "base64url-koodattu merkkijono",
+  json_string: "JSON-merkkijono",
+  e164: "E.164-luku",
+  jwt: "JWT",
+  template_literal: "templaattimerkkijono",
+};
+
+const InOrigin: { [k in string & {}]?: string } = {
+  record: "tietueessa",
+  map: "hakemistossa",
+  set: "joukossa",
+};
+
+const error: errors.$ZodErrorMap = (issue) => {
+  switch (issue.code) {
+    case "invalid_type":
+      return `Virheellinen tyyppi: täytyy olla ${getTypeName(issue.expected)}, oli ${parsedType(issue.input)}`;
```

Comment:
Pallottelen vaihtoehtoisella sanamuodolla
```suggestion
      return `Virheellinen tyyppi: odotettiin ${getTypeName(issue.expected)}, oli ${parsedType(issue.input)}`;
```

Joka tapauksessa tämä "täytyy olla" IMO huomattavasti parempi kuin alkuperäinen `"pitäisi olla"` 👍

Reply:
There seems to also be some older `fi.json` in
[aiji42//zod-i18n](https://github.com/aiji42/zod-i18n/blob/main/packages/core/locales/fi/zod.json) library which has same error translated as:

> `"Odotettiin arvon olevan {{expected}}, saatiin {{received}}"`

But yeah, not sure, food for thought.

Reply:
Tbh harkitsen vahvasti sitä, että heivais kuitenkin nää datatyyppien käännökset kokonaan pois tästä kohtaa, kun näitä JS datatyyppejä on niin mahdotonta saada mielekkäästi ja tyhjentävästi hoidettua. Voisi vaan olla esim tuo:
```
"Virheellinen tyyppi: odotettiin {{issue.expected}}, saatiin {{parsedType(issue.input)}}"
```
jossa tuo `parsedType` vaan palauttaa saman kuin tuo englanninkielinen versio.